File: toolchains/node/src/trie.js
const { TrieNode } = require('./trieNode');

/**
 * Generic trie (data-structures toolchain).
 */
class Trie {
  constructor(root = null) {
    this._root = root;
  }

  get root() {
    return this._root;
  }

  isEmpty() {
    return this._root == null;
  }

  size() {
    if (this._root == null) return 0;
    return this._countNodes(this._root);
  }

  _countNodes(node) {
    let count = 1;
    for (const child of node.children) {
      count += this._countNodes(child);
    }
    return count;
  }
}

module.exports = { Trie, TrieNode };

File: toolchains/csharp/Trie.cs
namespace LangChain;

/// <summary>Generic trie (LangChain namespace: chain/trie structures).</summary>
public class Trie<T>
{
    public TrieNode<T>? Root { get; set; }

    public Trie(TrieNode<T>? root = null) => Root = root;

    public bool IsEmpty => Root == null;

    public int Size => Root == null ? 0 : CountNodes(Root);

    static int CountNodes(TrieNode<T> node)
    {
        int count = 1;
        foreach (var child in node.Children)
            count += CountNodes(child);
        return count;
    }
}
